File: models/plant.go
package models

import (
	"encoding/json"
	"time"

	"github.com/gobuffalo/pop/v6"
	"github.com/gobuffalo/validate/v3"
	"github.com/gobuffalo/validate/v3/validators"
	"github.com/gofrs/uuid"
)

// Plant is used by pop to map your plants database table to your go code.
type Plant struct {
	ID            uuid.UUID `json:"id" db:"id"`
	Name          string    `json:"name" db:"name"`
	Germinated    bool      `json:"germinated" db:"germinated"`
	DaysToHarvest int       `json:"days_to_harvest" db:"days_to_harvest"`
	PlantCount    int       `json:"plant_count" db:"plant_count"`
	DatePlanted   time.Time `json:"date_planted" db:"date_planted"`
	DateGerminated time.Time `json:"date_germinated" db:"date_germinated"`
	GardenID      uuid.UUID `json:"garden_id" db:"garden_id"`
	Garden        *Garden   `json:"Garden,omitempty" belongs_to:"garden"`
	PlantTags     Tags      `json:"plant_tags,omitempty" many_to_many:"plants_tags"`
	Journals      []Journal `json:"journals,omitempty" has_many:"journals"`
	WaterSchedules WaterSchedule `has_one:"water_schedules"`
	AccountID     uuid.UUID `json:"account_id" db:"account_id"`
	Account       *Account  `json:"Account,omitempty" belongs_to:"accounts"`
	CreatedAt     time.Time `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time `json:"updated_at" db:"updated_at"`
}

func (p Plant) PlantName() string {
	return p.Name
}

func (p Plant) GetJournals(tx *pop.Connection) error {
	var journals []Journal
	err := tx.Where("plant_id = ?", p.ID).All(&journals)
	if err != nil {
		return err
	}

	p.Journals = journals
	return nil
}

func (p Plant) SelectLabel() string {
	return p.Name
}

func (p Plant) SelectValue() interface{} {
	return p.ID
}
// String is not required by pop and may be deleted
func (p Plant) String() string {
	jp, _ := json.Marshal(p)
	return string(jp)
}

// Plants is not required by pop and may be deleted
type Plants []Plant

// String is not required by pop and may be deleted
func (p Plants) String() string {
	jp, _ := json.Marshal(p)
	return string(jp)
}

// Validate gets run every time you call a "pop.Validate*" (pop.ValidateAndSave, pop.ValidateAndCreate, pop.ValidateAndUpdate) method.
// This method is not required and may be deleted.
func (p *Plant) Validate(tx *pop.Connection) (*validate.Errors, error) {
	return validate.Validate(
		&validators.StringIsPresent{Field: p.Name, Name: "Name"},
		&validators.IntIsPresent{Field: p.DaysToHarvest, Name: "DaysToHarvest"},
	), nil
}

// ValidateCreate gets run every time you call "pop.ValidateAndCreate" method.
// This method is not required and may be deleted.
func (p *Plant) ValidateCreate(tx *pop.Connection) (*validate.Errors, error) {
	return validate.NewErrors(), nil
}

// ValidateUpdate gets run every time you call "pop.ValidateAndUpdate" method.
// This method is not required and may be deleted.
func (p *Plant) ValidateUpdate(tx *pop.Connection) (*validate.Errors, error) {
	return validate.NewErrors(), nil
}

File: frontend/lib/model/plant.dart
import 'journal.dart';
import 'apis/journal_api.dart';
import 'ws.dart';
import 'apis/ws_api.dart';
import 'tag.dart';
import 'apis/tag_api.dart';

class Plant {
  final String id;
  final String name;
  final bool germinated;
  final int days_to_harvest;
  final int plant_count;
  final DateTime? date_planted;
  final DateTime? date_germinated;
  final String garden_id;
  final List<Tag>? plantTags;
  final List<Journal>? journals;
  final WaterSchedule? waterSchedules;
  final String account_id;
  final DateTime createdAt;
  final DateTime updatedAt;

  Plant(
      {required this.id,
      required this.name,
      required this.germinated,
      required this.days_to_harvest,
      required this.plant_count,
      this.date_planted,
      this.date_germinated,
      required this.garden_id,
      this.plantTags,
      this.journals,
      this.waterSchedules,
      required this.account_id,
      required this.createdAt,
      required this.updatedAt});

  factory Plant.fromJson(Map<String, dynamic> json) {
    bool parseBool(dynamic value) {
      if (value is bool) return value;
      if (value is int) return value == 1;
      throw ArgumentError('Invalid boolean value: $value');
    }

    return Plant(
      id: json['id'],
      name: json['name'],
      germinated: parseBool(json['germinated']),
      days_to_harvest: json['days_to_harvest'],
      plant_count: json['plant_count'],
      date_planted: json['date_planted'] != null && json['date_planted'] != ''
          ? DateTime.parse(json['date_planted'])
          : null,
      date_germinated:
          json['date_germinated'] != null && json['date_germinated'] != ''
              ? DateTime.parse(json['date_germinated'])
              : null,
      garden_id: json['garden_id'],
      plantTags: json['plantTags'] != null
          ? List<Tag>.from(json['plantTags'].map((x) => Tag.fromJson(x)))
          : null,
      journals: json['journals'] != null
          ? List<Journal>.from(json['journals'].map((x) => Journal.fromJson(x)))
          : null,
      waterSchedules: json['waterSchedules'] != null
          ? WaterSchedule.fromJson(json['waterSchedules'])
          : null,
      account_id: json['account_id'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson({bool forSqlLite = false}) {
    dynamic serializeBool(bool value) {
      return forSqlLite ? (value ? 1 : 0) : (value ? true : false);
    }

    return {
      'id': id,
      'name': name,
      'germinated': serializeBool(germinated),
      'days_to_harvest': days_to_harvest,
      'plant_count': plant_count,
      'date_planted': date_planted?.toIso8601String(),
      'date_germinated': date_germinated?.toIso8601String(),
      'garden_id': garden_id,
      'PlantTags': plantTags?.map((x) => x.toJson()).toList(),
      'journals': journals?.map((x) => x.toJson()).toList(),
      'waterSchedules': waterSchedules?.toJson(),
      'account_id': account_id,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
